### ISSUE TITLE
[FLI-405] List/Read API Endpoints for all resources

### ISSUE THREAD

**Issue by markphelps** [flipt-io/flipt#1730]
### Problem

Is your feature request related to a problem? Please describe:

Currently, some of our resources such as:

- [Variants](https://www.flipt.io/docs/concepts#variants)
- [Constraints](https://www.flipt.io/docs/concepts#constraints)
- [Distributions](https://www.flipt.io/docs/concepts#distributions)

Are only available to be read from their 'parent' resources ie:

- [GetFlag](https://www.flipt.io/docs/reference/flags/get-flag) returns a Flag along with a list of Variants
- [GetSegment](https://www.flipt.io/docs/reference/segments/get-segment) returns a Segment along with a list of Constraints
- [GetRule](https://www.flipt.io/docs/reference/rules/get-rule) returns a Rule along with a list of Distributions

![CleanShot 2023-06-09 at 09 35 30](https://github.com/flipt-io/flipt/assets/209477/16d8296c-8ec3-4cad-a61a-be5bc4011880)

This means there isn't a way to access these sub-resources without going through the parent.

This has a few downsides including:

- There is **no way to paginate** a list of variants for example. If a user has 1000 variants for a flag, they get all of them in the Get/List Flag response.
- It makes the API inconsistent and not very RESTful

### Ideal Solution

Describe the solution you'd like:

Ideally, each resource will have the following API endpoints where it makes sense:

- GetX
- ListX (with pagination)

This means we would add the following endpoints to our API:

- GetVariant
- ListVariants
- GetConstraint
- ListConstraints
- GetDistribution
- ListDistributions

#### Deprecation

We should also deprecate the fields in the `Get/ListFlag(s)`, `Get/ListSegment(s)`, `Get/ListRule(s)` responses. Meaning we should deprecate:

- `variants` from `Flag` response
- `constraints` from `Segment` response
-  `distributions` from `Rule` response

We should encourage callers to instead get these resources from the new API endpoints listed above.

### Additional Context

This will have far-reaching effects to various parts of the system including:

- The need to update the UI as it depends on these resources being part of their parent at the moment
- Export functionality (also dependent on nesting)

These changes can be made over time though.

#### How to Implement

If someone wanted to take a crack at implementing the above new API endpoints, here are some rough instructions on how to get started:

It all starts here with the proto service RPC definition:
https://github.com/flipt-io/flipt/blob/76010ad5436477823bb70c30755a0c8168a701b5/rpc/flipt/flipt.proto#L376-L409
Followed by an equivalent entry in here (this make the HTTP endpoints appear):
https://github.com/flipt-io/flipt/blob/main/rpc/flipt/flipt.yaml

We could e.g. add a ListVariants method to these.

The following then generates the types we need to fill in: `mage go:proto`

Flipt has a couple of layers to fill in; a server and a storage layer implementation.

Here is the ListFlags server layer: https://github.com/flipt-io/flipt/blob/main/internal/server/flag.go#L44-L85
Then the equivalent storage layer implementation: https://github.com/flipt-io/flipt/blob/main/internal/storage/sql/common/flag.go#L137-L248

We would add ListVariants methods to both of those. The server implementation would call the storage one.
That would be the crux of it, plus we would add also some tests.

<sub>[FLI-405](https://linear.app/flipt/issue/FLI-405/listread-api-endpoints-for-all-resources)</sub>

**Comment by GeorgeMac** [comment]
I dug into this issue a little and reflected on whether or not we really need them for Flipt, while investigating some improvements to our FS backend work. The TL;DR is that I don't think we need them and that the problem of pagination is not a real problem for anyone. So I would like to avoid adding and having to support them if at all possible.

The ask came from @jondavidjohn to support his work building a Terraform provider (❤️) and I would like to propose we explore some alternative ways to shape the provider to avoid needing to add these endpoints. I think the result can be more appropriate for our existing domain model.

TL;DR is I think we should forgo adding these and take a step back and reflect on the original problem a bit more.

## Assumptions

Here are assumptions I made based on our Discord discussion for the want here (correct me if they are wrong @jondavidjohn):

The need for these endpoints arrises from the fact that the provider being built treats the following concepts in Flipt as top-level Terraform resources:

- Namespace
- Flag
- Variant
- Rule
- Distribution
- Constraint
- Segment

## Flipts existing resources

Flipt really only has three valuable, addressable top-level resources:
- Namespace
- Flag
- Segment

This is purely from the needs within Flipt itself. Namespaces provide the grouping construct. Evaluations (the core offering of Flipt) are done within the context of a single Flag. Segments provide a re-usable way of declaring targetting on evaluation contexts.

Everything else belongs to Flag or Segment and support them doing their jobs. They don't make sense as concepts outside of these top-level resources. Our import and export format, while it has its own UX problems, does exemplify the heirarchy of Flipt resources well and is why it was possible to use as the FS backend declarative format:

https://github.com/flipt-io/flipt/blob/e1b4ea056213e4757be1c93604699ee56934a264/internal/ext/testdata/export.yml#L1-L47

I can see why, given we supported each of the sub-resources as top-level Terraform resources, we would need to support listing and getting them. Flipts API implies that each of these resources can be handled in isolation as we support some getters and CUD operations for things like Variant, Rule, Distribution and Constraint.

For the pagination argument I would really just say that we've never had an ask for this in Flipts lifetime.
I would very much question someones setup if they said they had more than a handful of variants, constraints, rules and so on. The complexity of doing this for a real-world usecase would become untenable.

## Proposal for Terraform Provider

Could we simply only expose `Namespace`, `Flag` and `Segment` as top-level Terraform resources?
Then use some of the aggregate schema types for the resource attributes to declare variants, rules and so on?

e.g. (I am just sketching here) the declarative format would allow you to do something like:
```hcl
resource "namespace" "production" {}

resource "flag" "foo" {
  namespace = "production"
  enabled = true
  variants = [
    {
      key = "red"
    },
    {
      key = "blue"
    },
  ]
  rules = [
    {
      segment = "segment_one"
      distributions = [
        {
          variant = "red"
          rollout = 50.0
        }
      ]
    },
  ]
}

resource "segment" "segment_one" {
  match_type = "ANY_MATCH_TYPE"
  constraints = [
    {
      comparison_type = "STRING_COMPARISON_TYPE"
      property = "name"
      operator = "eq"
      value = "user_one"
    },
  ]
}
```

> I read somewhere that there is the concept of a `schema.ListNestedAttributes` that might be appropriate here for these types.

I would be happy to help figure out how this would be possible.
I could even break this out into a Terraform provider discussion / design doc / issue.

Our internal import and export functionality effectively does this against our API today.
It walks that yaml format I linked earlier and exercises it against the Flipt API:
https://github.com/flipt-io/flipt/blob/main/internal/ext/importer.go#L58-L252

## Potential Future Changes for Flipt

This one in particular is for @markphelps to consider:
An additional thing we could do is actually explore a new set of declarative endpoints for writing Flipt state.
I think these could be valuable in general going forward for Terraform, server side import and the filesystem backends.

This could just be two new endpoints to start off with. In particular, `PUT flag` and `PUT segment` taking all the state needed for a flag or segment, which would just bulk overwrite the resource referenced.

I reckon this would work well for most usecases. If we needed to support smaller / finer edits, then we could support `PATCH flag` and `PATH segment` which would contain a bunch of optional commutative operations (replace name, append rule and so on).

I think we could do a lot with this API and it might allow us to hedge our bets for supporting GitOps style workloads.
We could leverage this to simplify building a k8s controller with CRDs for Flipt resources.
